pkg/providers/instanceprofile/instanceprofile.go

```diff
@@ -0,0 +1,154 @@
+		}
+		instanceProfile = o.InstanceProfile
+	} else {
+		instanceProfile = out.InstanceProfile
+	}
+	if len(instanceProfile.Roles) == 1 {
+		return profileName, nil
+	}
+	if _, err = p.iamapi.AddRoleToInstanceProfileWithContext(ctx, &iam.AddRoleToInstanceProfileInput{
+		InstanceProfileName: aws.String(profileName),
+		RoleName:            aws.String(nodeClass.Spec.Role),
+	}); err != nil {
+		return "", fmt.Errorf("adding role %q to instance profile %q, %w", nodeClass.Spec.Role, profileName, err)
+	}
+	p.cache.SetDefault(string(nodeClass.UID), nil)
+	return profileName, nil
+}
+
+func (p *Provider) AssociatedInstances(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) ([]string, error) {
+	profileName := p.GetProfileName(ctx, nodeClass)
+
+	// Get all instances that are using our instance profile name and are not yet terminated
+	var ids []string
+	if err := p.ec2api.DescribeInstancesPagesWithContext(ctx, &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("iam-instance-profile.name"),
+				Values: aws.StringSlice([]string{profileName}),
+			},
+			instanceStateFilter,
+		},
+	}, func(page *ec2.DescribeInstancesOutput, _ bool) bool {
+		for _, res := range page.Reservations {
+			ids = append(ids, lo.Map(res.Instances, func(i *ec2.Instance, _ int) string {
+				return aws.StringValue(i.InstanceId)
+			})...)
+		}
+		return true
+	}); err != nil {
+		return nil, fmt.Errorf("getting associated instances for instance profile %q, %w", profileName, err)
+	}
+	return ids, nil
+}
+
+func (p *Provider) Delete(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) error {
+	profileName := p.GetProfileName(ctx, nodeClass)
+	out, err := p.iamapi.GetInstanceProfileWithContext(ctx, &iam.GetInstanceProfileInput{
+		InstanceProfileName: aws.String(profileName),
+	})
+	if err != nil {
+		return awserrors.IgnoreNotFound(fmt.Errorf("getting instance profile %q, %w", profileName, err))
+	}
+	if len(out.InstanceProfile.Roles) > 0 {
+		for _, role := range out.InstanceProfile.Roles {
+			if _, err = p.iamapi.RemoveRoleFromInstanceProfileWithContext(ctx, &iam.RemoveRoleFromInstanceProfileInput{
+				InstanceProfileName: aws.String(profileName),
+				RoleName:            role.RoleName,
+			}); err != nil {
+				return fmt.Errorf("removing role %q from instance profile %q, %w", aws.StringValue(role.RoleName), profileName, err)
+			}
+		}
+	}
+	if _, err = p.iamapi.DeleteInstanceProfileWithContext(ctx, &iam.DeleteInstanceProfileInput{
+		InstanceProfileName: aws.String(profileName),
+	}); err != nil {
+		return awserrors.IgnoreNotFound(fmt.Errorf("deleting instance profile %q, %w", profileName, err))
+	}
+	return nil
+}
+
+func (p *Provider) GetProfileName(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) string {
+	return fmt.Sprintf("%s/%d", settings.FromContext(ctx).ClusterName, lo.Must(hashstructure.Hash(fmt.Sprintf("%s/%s", nodeClass.Name, p.region), hashstructure.FormatV2, &hashstructure.HashOptions{})))
```

Comment:
Does it make sense to not hash this? I can see this being a nice observability pro to know which node class each Instance profile refers to easily.

Reply:
So if we don't hash it, then we have to truncate it, at which point we can't guarantee the name is sufficiently random. This is part of the reason that we add the tags to the instance profile so that users can track down the associated instance profile if they need to.

Reply:
There's probably also room here down the line to do the async reconciliation of this instance profile, where we propagate the name into the status, at which point that would be the observability piece that we are looking for here. But the real limitation here is:
1. The EKS cluster name limit is 100 chars
2. The Node Class name limit is 64 chars
3. The Instance profile name limit is 128 chars

We necessarily can't capture all the information here. We could consider _just_ using the NodeClass UUID as another option.